extension/src/prototypes/ui-components/TimelineBar.tsx

```diff
@@ -0,0 +1,324 @@
+    date.setMinutes(0);
+    date.setSeconds(0);
+    date.setMilliseconds(0);
+
+    // Need to check the start since we consider timezone in hours
+    if (date.getTime() < duration.msStart) date.setFullYear(date.getFullYear() + 1);
+    while (date.getTime() < duration.msEnd) {
+      const ms = date.getTime();
+      ticks.push({
+        left: ((ms - duration.msStart) / duration.msDuration) * 100,
+        ms,
+      });
+      date.setFullYear(date.getFullYear() + 1);
+    }
+  } else {
+    const start = msStep - (duration.msStart % msStep);
+    for (let t = duration.msStart + start; t <= duration.msEnd; t += msStep) {
+      ticks.push({ left: ((t - duration.msStart) / duration.msDuration) * 100, ms: t });
+    }
+  }
+  return ticks;
+};
+
+type LabelProps = {
+  date: Date;
+  left: number;
+  timezone: string;
+  level?: number;
+};
+const Label: React.FC<LabelProps> = ({ date, left, level }) => {
+  const [dateString, timeString] = formatDate(date, level && level >= 7 ? "year" : "normal");
+  return (
+    <StyledLabel left={left}>
+      <DateLabel>{dateString}</DateLabel>
+      <TimeLabel>{timeString}</TimeLabel>
+    </StyledLabel>
+  );
+};
+
+const TimelineBar: React.FC<TimelineBarProps> = ({
+  startDate,
+  endDate,
+  currentDate,
+  timezone,
+  onChange,
+}) => {
+  const [fullWidth, setFullWidth] = useState(defaultFullWidth);
+  const [maxLevel, setMaxLevel] = useState(0);
+
+  const duration: Duration = useMemo(() => {
+    const msStart = getDateWithTimezone(startDate, timezone).getTime();
+    const msEnd = getDateWithTimezone(endDate, timezone).getTime();
+    const msDuration = msEnd - msStart;
+    return {
+      msStart,
+      msEnd,
+      msDuration,
+    };
+  }, [startDate, endDate, timezone]);
+
+  const trackWrapperRef = useRef<HTMLDivElement>(null);
+  useEffect(() => {
+    const resizeObserver = new ResizeObserver(entries => {
+      const [entry] = entries;
+      let width: number | undefined;
+      if (entry.borderBoxSize?.length > 0) {
+        const borderBoxSize = Array.isArray(entry.borderBoxSize)
+          ? entry.borderBoxSize[0]
+          : entry.borderBoxSize;
+        width = borderBoxSize.inlineSize;
+      } else if (entry.contentRect) {
+        width = entry.contentRect.width;
+      } else {
+        width = trackWrapperRef.current?.clientWidth;
+      }
+      setFullWidth(width ?? defaultFullWidth);
+    });
+    if (trackWrapperRef.current) {
+      resizeObserver.observe(trackWrapperRef.current);
+    }
+    return () => {
+      resizeObserver.disconnect();
+    };
+  }, [trackWrapperRef]);
+
+  const tickGroups = useMemo(() => {
+    let innerMaxLevel = 0;
+    const tg = tickGroupSettings.map((type, i) => {
+      const ticks = getTicks(duration, fullWidth, type.ms, type.name, timezone);
+      if (ticks.length > 0 && i > innerMaxLevel) innerMaxLevel = i;
+      return {
+        ticks,
+        name: type.name,
+      };
+    });
+    setMaxLevel(innerMaxLevel);
```

Comment:
Why do we need to re-render in here? Does it not work just returning the `innerMaxLevel`?

Reply:
Yes this looks much better. I just didn't aware return array when using useMemo. Updated.